File: src/Marsen.NetCore.Dojo/Classes/Joey/Tennis/StateV2/NormalState.cs
namespace Marsen.NetCore.Dojo.Classes.Joey.Tennis.StateV2;

public class NormalState : State
{
    public override string Score()
    {
        return $"{ScoreLookup[Game.ServerPoint]} {ScoreLookup[Game.ReceiverPoint]}";
    }

    protected override void ChangeState()
    {
        State state = IsReadyToWin() ? new WinState() : new NormalState();

        if (IsSame()) state = Game.ServerPoint >= 3 ? new DeuceState() : new SameState();

        Game.ChangeState(state);
    }

    private bool IsReadyToWin()
    {
        return Game.ServerPoint > 3 || Game.ReceiverPoint > 3;
    }
}

File: src/Marsen.TypeScript.Dojo/src/Kata/NormalState.ts
import { WinState } from "./WinState";
import { State } from "./State";
import { SameState } from "./SameState";
import { DeuceState } from "./DeuceState";

export class NormalState extends State {
    ChangeState() {

        if (this.IsSame()) {
            this.Context.State = this.Context.ServerPoint < 3 ? new SameState() : new DeuceState();        
        }

        if (this.Context.ServerPoint > 3 || this.Context.ReceiverPoint > 3) {
            this.Context.State = new WinState();
        }
    }


    Score(): string {
        return `${this.ScoreLookup.get(this.Context.ServerPoint)} ${this.ScoreLookup.get(this.Context.ReceiverPoint)}`;
    }
}
